src/store_client.cc

```diff
@@ -499,58 +517,126 @@ store_client::fileRead()
+    assert(flags.disk_io_pending);
     flags.disk_io_pending = false;
     assert(_callback.pending());
-    debugs(90, 3, "storeClientReadBody: len " << len << "");
+    Assure(parsingBuffer);
+    debugs(90, 3, "got " << lastIoResult << " using " << *parsingBuffer);
 
-    if (len < 0)
+    if (lastIoResult < 0)
         return fail();
 
-    const auto rep = entry->mem_obj ? &entry->mem().baseReply() : nullptr;
-    if (copyInto.offset == 0 && len > 0 && rep && rep->sline.status() == Http::scNone) {
-        /* Our structure ! */
-        if (!entry->mem_obj->adjustableBaseReply().parseCharBuf(copyInto.data, headersEnd(copyInto.data, len))) {
-            debugs(90, DBG_CRITICAL, "ERROR: Could not parse headers from on disk object");
-        }
+    if (!lastIoResult) {
+        if (answeredOnce())
+            return noteEof();
+
+        debugs(90, DBG_CRITICAL, "ERROR: Truncated HTTP headers in on-disk object");
+        return fail();
     }
 
-    if (len > 0 && rep && entry->mem_obj->inmem_lo == 0 && entry->objectLen() <= (int64_t)Config.Store.maxInMemObjSize && Config.onoff.memory_cache_disk) {
-        storeGetMemSpace(len);
+    assert(lastRead_.data == buf);
+    lastRead_.length = lastIoResult;
+
+    parsingBuffer->appended(buf, lastIoResult);
+
+    maybeWriteFromDiskToMemory(lastRead_);
+    handleBodyFromDisk();
+}
+
+/// de-serializes HTTP response (partially) read from disk storage
+void
+store_client::handleBodyFromDisk()
+{
+    // We cannot de-serialize on-disk HTTP response without MemObject because
+    // without MemObject::swap_hdr_sz we cannot know where that response starts.
+    Assure(entry->mem_obj);
+    Assure(entry->mem_obj->swap_hdr_sz > 0);
+
+    if (!answeredOnce()) {
+        // All on-disk responses have HTTP headers. First disk body read(s)
+        // include HTTP headers that we must parse (if needed) and skip.
+        const auto haveHttpHeaders = entry->mem_obj->baseReply().pstate == Http::Message::psParsed;
+        if (!haveHttpHeaders && !parseHttpHeadersFromDisk())
+            return;
+        skipHttpHeadersFromDisk();
+    }
+
+    // no callback(): we handled negative/zero lastIoResult ourselves
+    noteNews();
+}
+
+/// Adds HTTP response data loaded from disk to the memory cache (if
+/// needed/possible). The given part may contain portions of HTTP response
+/// headers and/or HTTP response body.
+void
+store_client::maybeWriteFromDiskToMemory(const StoreIOBuffer &httpResponsePart)
+{
+    // XXX: Reject [memory-]uncachable/unshareable responses instead of assuming
+    // that any HTTP response loaded from disk should be written to MemObject's
+    // data_hdr (and that it may purge already cached entries). Cachability
+    // decision(s) should be made outside (and obeyed by) this low-level code.
+    if (httpResponsePart.length && entry->mem_obj->inmem_lo == 0 && entry->objectLen() <= (int64_t)Config.Store.maxInMemObjSize && Config.onoff.memory_cache_disk) {
```

Comment:
Do you agree with this XXX? This question has a low priority.

Reply:
>  any HTTP response loaded from disk should be written to MemObject's data_hdr

This condition does not allow 'any' response, since it checks at least two parameters, Config.Store.maxInMemObjSize and Config.onoff.memory_cache_disk. I agree that not all memory-cache-limiting conditions are checked here, for example, Config.memMaxSize is missed.

Reply:
I will rephrase in hope to clarify that "any" or "should" or square brackets: The point of this XXX is very broad -- we do not know whether this response is cachable at all (not just memory-cachable but cachable in general). For example, the response may have been marked for release while we were loading it from disk. Or perhaps the cache admission rules in squid.conf have changed since the object got stored on disk.

Somebody somewhere needs to make the (complex) decision to mem-cache this response. Here, we should be obeying that earlier decision (instead of poorly making our own decision every time new bytes come from disk). There might be some exceptional conditions to double check here (to change our mind and stop caching), but that is not what this code is currently checking/doing.

Reply:
Done in 235a8ee.